stan/math/rev/functor/reduce_sum.hpp

```diff
@@ -0,0 +1,523 @@
+#ifndef STAN_MATH_REV_SCAL_FUNCTOR_REDUCE_SUM_HPP
+#define STAN_MATH_REV_SCAL_FUNCTOR_REDUCE_SUM_HPP
+
+#include <stan/math/prim/meta.hpp>
+#include <stan/math/prim/functor.hpp>
+#include <stan/math/rev/fun.hpp>
+#include <tbb/task_arena.h>
+#include <tbb/parallel_reduce.h>
+#include <tbb/blocked_range.h>
+
+#include <iostream>
+#include <iterator>
+#include <tuple>
+#include <algorithm>
+#include <vector>
+
+namespace stan {
+namespace math {
+namespace internal {
+
+/**
+ * Var specialization of implimentation called by reduce `reduce_sum`.
+ * @tparam ReduceFunction An type with a valid `operator()`
+ * @tparam ReturnType The return type of the call to
+ * `ReduceFunction::operator()`
+ * @tparam Vec type of the first input argument with an `operator[]`
+ * @tparam Args Parameter pack holding the types of the object send to
+ * `ReduceFunction::operator()`
+ */
+template <typename ReduceFunction, typename ReturnType, typename Vec,
+          typename... Args>
+struct reduce_sum_impl<ReduceFunction, require_var_t<ReturnType>, ReturnType,
+                       Vec, Args...> {
+  /**
+   * Internal object used in `tbb::parallel_reduce`
+   * @note see link [here](https://tinyurl.com/vp7xw2t) for requirements.
+   */
+  struct recursive_reducer {
+    size_t per_job_sliced_terms_;
+    size_t num_shared_terms_;  // Number of terms shared across threads
+    double* sliced_partials_;  // Points to adjoints of the partial calculations
+    Vec vmapped_;
+    std::ostream* msgs_;
+    std::tuple<Args...> args_tuple_;
+    double sum_{0.0};
+    Eigen::VectorXd args_adjoints_{0};
+
+    template <typename VecT, typename... ArgsT>
+    recursive_reducer(size_t per_job_sliced_terms, size_t num_shared_terms,
+                      double* sliced_partials, VecT&& vmapped,
+                      std::ostream* msgs, ArgsT&&... args)
+        : per_job_sliced_terms_(per_job_sliced_terms),
+          num_shared_terms_(num_shared_terms),
+          sliced_partials_(sliced_partials),
+          vmapped_(std::forward<VecT>(vmapped)),
+          msgs_(msgs),
+          args_tuple_(std::forward<ArgsT>(args)...) {}
```

Comment:
Is it possible that we create a copy of the shared arguments in args_tuple for every instance of recursive_reducer here? Or is the `std::tuple<Args...>` really just a reference holder for the arguments which have been passed. So does the tuple store the data or just the references to things?

(sorry for these Qs... that's new C++ for me)

Reply:
> Is it possible that we create a copy of the shared arguments in args_tuple for every instance of recursive_reducer here? Or is the std::tuple<Args...> really just a reference holder for the arguments which have been passed. So does the tuple store the data or just the references to things?

TL;DR: No copies! It just stores references

So first let's look at how this is called. 

```cpp
template <typename ReduceFunction, typename Vec,
          typename = require_vector_like_t<Vec>, typename... Args>
auto reduce_sum(Vec&& vmapped, std::size_t grainsize, std::ostream* msgs,
                Args&&... args) {
  using return_type = return_type_t<Vec, Args...>;
  return internal::reduce_sum_impl<ReduceFunction, void, return_type, Vec,
                                   Args...>()(
      std::forward<Vec>(vmapped), grainsize, msgs, std::forward<Args>(args)...);
}
```

The top level class's templates are passed in as perfect forwarding references which preserve their value type. The two ampersands after the argument `Args&&` are normally used to define rvalues (references to temporary objects). However, because C++ is C++ they decided that inside of function signatures, when the argument is purely a template type with no additional attributes (like const) and has an `&&`, then the argument accepts any reference type such as lvalue `Arg&`, const lvalue `const Arg&` and rvalue `const Arg&&`. These are commonly called "Universal Reference Arguments", you can read more about them [here](https://isocpp.org/blog/2012/11/universal-references-in-c11-scott-meyers)

So what all that means is that when we define `reduce_sum_impl` with those templates _the value types are preserved_, so `Vec` here is actually one of `Vec&`, `const Vec&`, or if it's a temporary then `Vec&&`. You can see this at the below.

https://godbolt.org/z/q7EzM_

`args_tuple_` is defined as `std::tuple<Args...>`, Because `Args` comes from `reduce_sum`, `Args` is a universal reference parameter pack holding other reference types. When expanded that can hold references like `std::vector<double>&, Eigen::MatrixXd&&, const std::vector<var>&` etc. So everything in `args_tuple_` will be a reference.

`recursive_reducer` is called from within `reduce_sum_impl::operator()` as 

```cpp
    recursive_reducer worker(per_job_sliced_terms, num_shared_terms, partials,
                             vmapped, msgs, args...);
```

Where args is, `ArgsOp&&...` aka a universal reference parameter pack like the above. Since everything is coming in as references and being assigned to references there should not be any copies here.